File: forcast_desk/src/components/utility/expense_assumption/create_expense.js
import { createResource } from "frappe-ui"
import { selectedProject } from '@/components/utility/dashboard/projectService.js'

// Create a resource for expense document operations
const expenseResource = createResource({
  url: 'frappe.client.insert',
  makeParams({ doctype, doc }) {
    return {
      doc: {
        doctype,
        ...doc
      }
    }
  }
})

export async function createExpenseDocument({ year, month, expenses }) {
  // Get the currently selected project
  const currentProject = selectedProject.value
  
  if (!currentProject) {
    throw new Error('No project selected. Please select a project first.')
  }

  // Transform the expenses array to match the Expense Items table structure
  const expenseItems = expenses.map(expense => ({
    department: expense.department,
    department_location: expense.department_location,
    expense_name: expense.expense, 
    amount: expense.amount,
    // Add cost type if available
    ...(expense.costType && { cost_type: expense.costType }),
    // Add flag to indicate this is not a default expense (expenses from modal are always regular)
    is_default_expense: false,
  }))

  const doc = {
    year,
    month,
    project: currentProject.project_name, // Add project field
    expense_items: expenseItems,
  }


  try {
    const result = await expenseResource.submit({
      doctype: "Expense Assumptions",
      doc
    })
    
   //  // console.log('Document created successfully:', result)
    return { success: true, name: result.name }
  } catch (error) {
    console.error('Error creating expense document:', error)
    console.error('Error details:', {
      message: error.message,
      stack: error.stack,
      response: error.response
    })
    return { success: false, error }
  }
}

File: ex_forcast/api/create_expense.py
import frappe
import json

@frappe.whitelist(allow_guest=True)
def create_document(year, month, expenses, project=None):
    try:
        # Parse expenses if sent as JSON string
        if isinstance(expenses, str):
            expenses = json.loads(expenses)

        doc = frappe.new_doc("Expense Assumptions")
        doc.year = year
        doc.month = month
        doc.project = project  # Add project field

        # Normalize and append each expense row to match child doctype schema
        # Incoming row keys expected from frontend: department, department_location, expense, costType, amount/amountDisplay
        for item in expenses:
            department = item.get("department")
            department_location = item.get("department_location")
            expense_name = item.get("expense")
            cost_type = item.get("costType") or item.get("cost_type")
            # Prefer numeric amount; fallback to display string cleaned by frontend
            amount = item.get("amount")
            if amount is None and item.get("amountDisplay") is not None:
                # try to parse amountDisplay
                try:
                    amount = float(str(item.get("amountDisplay")).replace(",", ""))
                except (ValueError, TypeError):
                    amount = 0.0
            
            # Check if this is a default expense (you may need to implement logic to determine this)
            # For now, we'll assume all expenses from the modal are regular expenses
            is_default_expense = item.get("is_default_expense", False)
            
            # Create expense item data
            expense_item_data = {
                "department": department,
                "department_location": department_location,
                "amount": amount,
                "cost_type": cost_type
            }
            
            # Set the appropriate field based on whether it's a default expense
            if is_default_expense:
                expense_item_data["default_expense"] = expense_name
            else:
                expense_item_data["expense_name"] = expense_name
            
            doc.append("expense_items", expense_item_data)

        doc.insert()
        return {"success": True, "name": doc.name}

    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "create_expense_document failed")
        return {"success": False, "error": str(e)}


#  http://127.0.0.1:8000/api/v2/method/ex_forcast.api.create_expense.create_document